product_variant_exclusion/models/product.py

```diff
@@ -0,0 +1,174 @@
+                                product.unlink()
+                        except (psycopg2.Error, ValidationError):
+                            product.write({'active': False})
+                        break
+
+
+class ProductAttributeLine(models.Model):
+
+    _inherit = 'product.attribute.line'
+
+    @api.multi
+    def write(self, vals):
+        attr_values = vals.get('value_ids', False) and vals.get(
+            'value_ids')[0][2]
+        if attr_values and len(attr_values) <= 1:
+            for line in self:
+                for exclusion in line.product_tmpl_id.variant_exclusion_ids:
+                    if set(attr_values) <= set(
+                            exclusion.attribute_value_ids.ids):
+                        raise UserError(_(
+                            'You cannot leave only one attribute value on a '
+                            'line if this value is part of an exclusion'))
+        if self.value_ids.ids != attr_values:
+            new_values = self.env[
+                'product.attribute.value'].browse(attr_values)
+            exclusion_values = self.product_tmpl_id.variant_exclusion_ids\
+                .mapped('attribute_value_ids')
+            for new_val in new_values:
+                if new_val in exclusion_values:
+                    raise UserError(_('You cannot remove an attribute value '
+                                      'from an attribute line, if this value '
+                                      'is part of an exclusion.'))
+        return super(ProductAttributeLine, self).write(vals)
+
+    @api.multi
+    def unlink(self):
+        for line in self:
+            for val_id in line.value_ids.ids:
+                if val_id in line.product_tmpl_id.variant_exclusion_ids.mapped(
+                        'attribute_value_ids').ids:
+                    raise UserError(_(
+                        'You cannot delete an attribute line when '
+                        'one of its values is part of an exclusion'))
+        return super(ProductAttributeLine, self).unlink()
+
+
+class ProductVariantExclusion(models.Model):
+
+    _name = 'product.variant.exclusion'
+
+    product_tmpl_id = fields.Many2one('product.template', required=True)
+    attribute_value_ids = fields.Many2many('product.attribute.value',
+                                           required=True)
+
+    def _check_values_attributes_uniqueness(self, value_ids):
+        values = self.env['product.attribute.value'].browse(value_ids)
+        attributes = values.mapped('attribute_id')
+        if len(values) != len(attributes):
+            raise UserError(_('You cannot have multiple values from the '
+                              'same attribute in an exclusion.'))
+
+    @api.model
+    def create(self, vals):
+        value_ids = vals.get('attribute_value_ids')[0][2]
+        if len(value_ids) <= 1:
+            raise UserError(_(
+                'You cannot create an exclusion with only one value !'))
```

Comment:
For a lot of these errors, consider presenting them in the positive rather than the negative.

e.g Exclusions require more than one value.  Basically tell the user what they need to do, not what they don't need to.